extensions/websockets-next/server/deployment/src/main/java/io/quarkus/websockets/next/deployment/WebSocketServerProcessor.java

```diff
@@ -0,0 +1,657 @@
+import io.vertx.core.json.JsonObject;
+
+public class WebSocketServerProcessor {
+
+    static final String ENDPOINT_SUFFIX = "_WebSocketEndpoint";
+    static final String NESTED_SEPARATOR = "$_";
+
+    private static final Pattern PATH_PARAM_PATTERN = Pattern.compile("\\{[a-zA-Z0-9_]+\\}");
+
+    @BuildStep
+    FeatureBuildItem feature() {
+        return new FeatureBuildItem("websockets-next");
+    }
+
+    @BuildStep
+    BeanDefiningAnnotationBuildItem beanDefiningAnnotation() {
+        return new BeanDefiningAnnotationBuildItem(WebSocketDotNames.WEB_SOCKET, DotNames.SINGLETON);
+    }
+
+    @BuildStep
+    void unremovableBeans(BuildProducer<UnremovableBeanBuildItem> unremovableBeans) {
+        unremovableBeans.produce(UnremovableBeanBuildItem.beanTypes(TextMessageCodec.class));
+    }
+
+    @BuildStep
+    public void collectEndpoints(BeanArchiveIndexBuildItem beanArchiveIndex,
+            BeanDiscoveryFinishedBuildItem beanDiscoveryFinished,
+            BuildProducer<WebSocketEndpointBuildItem> endpoints) {
+
+        IndexView index = beanArchiveIndex.getIndex();
+        Set<String> paths = new HashSet<>();
+
+        for (BeanInfo bean : beanDiscoveryFinished.beanStream().classBeans()) {
+            ClassInfo beanClass = bean.getTarget().get().asClass();
+            AnnotationInstance webSocketAnnotation = beanClass.annotation(WebSocketDotNames.WEB_SOCKET);
+            if (webSocketAnnotation != null) {
+                String path = getPath(webSocketAnnotation.value().asString());
+                if (beanClass.nestingType() == NestingType.INNER) {
+                    // Sub-websocket - merge the path from the enclosing classes
+                    path = mergePath(getPathPrefix(index, beanClass.enclosingClass()), path);
+                }
+                if (!paths.add(path)) {
+                    throw new IllegalStateException("TODO multiple endpoints define the same path");
+                }
+                Callback onOpen = findCallback(beanArchiveIndex.getIndex(), beanClass, WebSocketDotNames.ON_OPEN,
+                        this::validateOnOpen);
+                Callback onMessage = findCallback(beanArchiveIndex.getIndex(), beanClass, WebSocketDotNames.ON_MESSAGE,
+                        this::validateOnMessage);
+                Callback onClose = findCallback(beanArchiveIndex.getIndex(), beanClass, WebSocketDotNames.ON_CLOSE,
+                        this::validateOnClose);
+                if (onOpen == null && onMessage == null) {
+                    throw new IllegalStateException(
+                            "TODO the endpoint must declare at least one method annotated with @OnMessage or @OnOpen");
+                }
+                endpoints.produce(new WebSocketEndpointBuildItem(bean, path, onOpen,
+                        onMessage, onClose));
+            }
+        }
+    }
+
+    @BuildStep
+    public void generateEndpoints(List<WebSocketEndpointBuildItem> endpoints,
```

Comment:
It's usually very useful to provide a sample of what this generates in the Javadoc of the method

Reply:
We can add a sample but the output is quite variable sooo :shrug:

Reply:
Yeah I know, but I have found in many many cases in the past that it gives someone a general idea of the bytecode generation is meant to do. The only real alternative is for readers of the code to do this on their own **every** time, which is suboptimal when causally looking at the code.